### ISSUE TITLE
Mismatch features number

### ISSUE THREAD

**Issue by mohamedalyzahran** [sgoldenlab/simba#304]
**Describe the bug**

**To Reproduce**
1. I am using your model of Mouse resident intruder classifier model 16bp on OSF 
2. And by trying to run the model I got the message of mismatched features number mentioning that my model is 490 while extracted features are 498
3. My videos trained by Deeblabcut were at 30fps
I tried to change the fps of the video using Simba to 25 (and also to 15) and I am getting the same message 
So, What can I do to solve this?

**Screenshots**

**Desktop (please complete the following information):**
 - OS: [win 10]
 - Python Version [e.g. 3.9.0]
 - Are you using anaconda? yes
 

**Additional context**

**Comment by sronilsson** [comment]
Hi @mohamedalyzahran - thanks for reporting, I think I know this one, and a fix - and there is no issue with your data, including fps, all on me..

A bit of background: Soon after we created the 16-body-part classifier on OSF in 2020-ish, we added 8 features to the 2-animal 16-body-part feature extraction class in SimBA. I can't remember why we added them ( think they were good for capturing rearing.). As others started using SimBA I ended up in a pickle - I couldn't remove the 8 features as people now needed them for their own models. In the lab we had switched to the 14-body-part model, so it also didn't make sense for me to spend time to create a new 16-body-part model and put it on OSF without the 8 features.

How about I point you to the 8 lines of code that computes them, and you can comment them out, and it should work?

**Comment by sronilsson** [comment]
A little bit hacky but...

You can find where simba is installed in your conda environt by typing `pip show simba-uw-tf-dev`:

<img width="861" alt="image" src="https://github.com/sgoldenlab/simba/assets/34761092/5d65678a-fa91-466d-aeac-bd3cde840a65">

From that location, go to `simba/feature_extractors/feature_extractor_16bp.py` and open it in a text editor.

The offending lines are 87-94, if you comment them out, restart SimBA, and let me know how it goes:

<img width="859" alt="image" src="https://github.com/sgoldenlab/simba/assets/34761092/184ef986-5eec-489b-a47d-bb14aa23781d">


These are the names of the 8 additional features you don’t want to calculate.

Mouse_1_Nose_to_lateral_left
Mouse_2_Nose_to_lateral_left
Mouse_1_Nose_to_lateral_right
Mouse_2_Nose_to_lateral_right
Mouse_1_Centroid_to_lateral_left
Mouse_2_Centroid_to_lateral_left
Mouse_1_Centroid_to_lateral_right
Mouse_2_Centroid_to_lateral_right

**Comment by sronilsson** [comment]
@mohamedalyzahran Yes some classifiers live on this OSF [gdrive expansion](https://osf.io/tmu6y/) as OSF started to limit data space. I can see three scikit random forest objects - the `w_operant` suffix I think means they have been trained on labelled data both from home cage and operant chamber. There is also a tracking model there, and labelled images used to train the model in DLC. If something is missing let me know and I can se if I can dig it up. 

<img width="537" alt="image" src="https://github.com/sgoldenlab/simba/assets/34761092/bc1e7c09-819b-4ed1-98a1-e15e04469f1d">

**Comment by mohamedalyzahran** [comment]
Hello 
I used the  your models of 16bp simba (the old(490) and the new(498) as well ) and were working but suddenly by extracting features and running the model I got this error however I didn't change anything and it was working with the same data
![image](https://github.com/sgoldenlab/simba/assets/123275892/91bedd83-3fb1-4328-999c-2b26f600999d)

**Comment by sronilsson** [comment]
Hi @mohamedalyzahran! What is happening, is that SimBA is looking for your body-parts coordinates columns inside the `project_folder/csv/features_extracted/MSWE_hM3Dq4.csv`,to drop them (we don't want them going into the classifier) before applying the classifier. For some reason, now SimBA can't find them (It is expecting the headers in that `BodyPartCoulmNotFoundWarning`), and proceeds. Next, when you come to the classifications, you have too many columns: you most likely still have the body-part coordinates still there (48 columns). 

So you know any reason for the body-parts changing names? 

If you open `MSWE_hM3Dq4.csc` and look at the initial column, can you check what the body-part columns are actually called?

**Comment by mohamedalyzahran** [comment]
the body parts are in the csv like before and the problem is that It was working perfectly with the same video and files, the only thing I did is just click extract feature more time and then suddenly it started to show this error i don't know why
![image](https://github.com/sgoldenlab/simba/assets/123275892/f31fd610-0e24-4f3e-a0f9-8cb6822aafc3)

**Comment by sronilsson** [comment]
Did you update SimBA just before it started playing up? 


Can you drop me a copy of the project_config.ini? Can you also drop me a copy here of your `project_folder/logs/measures/pose_configs/bp_names/project_bp_names.csv` file?

**Comment by mohamedalyzahran** [comment]
Also, how is the number of extracted features 546? it should be 498 or 490 if I used the old features extractor

**Comment by sronilsson** [comment]
548 is probably: all of your actual features (498) + the body-part columns (48) that SimBA fails to recognise as body-part columns and fails to drop

**Comment by sronilsson** [comment]
48 is 3 (x,y,p) x 16 (your body-part count) so would make sense

**Comment by mohamedalyzahran** [comment]
No, I didn't update it and it is the same error using two versions of simba and different projects as well 
[project_bp_names.csv](https://github.com/sgoldenlab/simba/files/13302235/project_bp_names.csv)
[project_config.zip](https://github.com/sgoldenlab/simba/files/13302241/project_config.zip)
If you mean that I have identified which is the resident and which is the intruder, yes I did

**Comment by mohamedalyzahran** [comment]
Another thing that I tried before I got this error is that I tried to apply user defined features using one of your uploaded script,
does this change anything in the default feature extraction?

**Comment by sronilsson** [comment]
Ah thank you, maybe. If its the default body-part config 16bp, SimBA doesn't assign animal names to the body-parts, it already knows who's who, but it does so with user-defined configs, so it likely changed your project_bp_names.csv, if you remove the `Intruder_` and `Resident_` prefixes from the project_bp_names.csv, save the file, and run it again, does it work?

**Comment by sronilsson** [comment]
So e.g., the first line reads `Ear_left_1` and so on

**Comment by mohamedalyzahran** [comment]
Awesome, It worked now by removing the Intruder_ and Resident_ prefixes, Thanks so much
A couple of questions if you don't mind: (1) if I try to apply user defined features, it gives me this error whileIam using the file extract_features_14bp_from_16bp.py or any other file.
(2) What is the best way to determine the threshold for each clasifier? is it running the integrative propability plot? and the minimum bout lenghth is there away?
(3) does the outlier corrections correct all body parts or only the ones I chose and as for location and movement critration should be specific number if so what is the most suitable?
(4) in your model of RI classifer like attack is measuring animal 1 am Iright? I mean how can I be sure that it is measuring the right animal nor the other?

thanks so much in advance

![image](https://github.com/sgoldenlab/simba/assets/123275892/9f070584-b7e5-44e9-b27f-ee50b39f7d90)

**Comment by sronilsson** [comment]
_(1)	if I try to apply user defined features, it gives me this error whileIam using the file extract_features_14bp_from_16bp.py or any other file._


So, the extract_features_14bp_from_16bp.py file is old, I am not sure it was written as a class, it may just be a function. How does it look if you e.g., run `simba/feature_extractors
/feature_extractor_16bp.py`.  It needs to be a class anyway to work, as documented [HERE]( https://github.com/sgoldenlab/simba/blob/master/docs/extractFeatures.md).


_(2) What is the best way to determine the threshold for each clasifier? is it running the integrative propability plot? and the minimum bout lenghth is there away?_

Yes I don’t have a better method, you could run precision-recall curves during training etc., but ideally you want to get the threshold for a different video, not used in training and testing.

_(3)	does the outlier corrections correct all body parts or only the ones I chose and as for location and movement critration should be specific number if so what is the most suitable?_

Outlier correction method is detailed [here]( https://github.com/sgoldenlab/simba/blob/master/misc/Outlier_settings.pdf) , the body-parts you choose is just a heurstic way of saying how big the animal is, it corrects all body-parts. Often I recommend just fixing the outliers by improving tracking model, or applying smoothing/interpolation. The criterion value can be a little tricky to figure out. Some people still use it though, there is even methods to apply different criterion and omit correction on some body-parts [here][https://simba-uw-tf-dev.readthedocs.io/en/latest/nb/advanced_outlier_correction.html], you can check this [issue]( https://github.com/sgoldenlab/simba/issues/306) from yesterday towards the end on how to evaluate it. 

_(3)	in your model of RI classifer like attack is measuring animal 1 am Iright? I mean how can I be sure that it is measuring the right animal nor the other?_

Not sure I understand but it expects this input: https://github.com/sgoldenlab/simba/blob/master/simba/pose_configurations/schematics/7.png

If you had a different pose-estimation input, where the first 8 body-parts actually belong to the C57, then there might be issues.

**Comment by mohamedalyzahran** [comment]
Thanks so much  for you reply.
Actually, I am using 2 c57 mice in each video (resident intruder 2 black
mice)
The annotations were done by dlc in which 1,2,3,4,5,6,7,8 body parts belong
to the resident animal and 9,10,11,12,13,14,15,16 belong to the intruder

So my question is that, are your classifiers(on osf  for 16bp) working in a
way that detect resident behavior in my case or I have to reverse animals
ID

**Comment by sronilsson** [comment]
I would probably stick to the way it is, but caveat I have never not tries using this cd1 c57 classifier on two c57s. 

But the classifier expects the first animal to be the resident, and the second animal to be the intruder - I wouldn't invert the data.

**Comment by mohamedalyzahran** [comment]
Do you mean by the first animal is just choosing animal1:Resident  and animal2: intruder and assign on each video whatever the annotations were in term of ordering (either the resident in green annotation or in pink) or first and second animals are about deeplabcut annotations orders (because by applying the DLC model it can't differentiate between resident and intruder)

![image](https://github.com/sgoldenlab/simba/assets/123275892/a8c377f1-d85d-4f1f-b7f5-634f6d6d5c06)
![image](https://github.com/sgoldenlab/simba/assets/123275892/2543e27b-b1f7-4dd2-a4dc-d72de419ad0f)
![image](https://github.com/sgoldenlab/simba/assets/123275892/ceeb7264-b7a8-4ab4-a138-1d43e7da4cc1)

**Comment by sronilsson** [comment]
Got it, sorry was thinking of non multi-animal tracking data. Yes, with maDLC, assign the resident as the first animal. /Simon

**Comment by mohamedalyzahran** [comment]
so it doesn't matter  the deeplabcut annotations order; I mean to have the resident in green annotations or in pink annotations(because in my case DLC doesn't differentiate animals so in some videos the resident is annotated in green dots and intruder in pink and in another video the reverse) 
So, Am I right? if I  assign animal1:resident to the resident animal on the video and it is not important in which color the annotations are?

**Comment by mohamedalyzahran** [comment]
it is clear now. thanks so much

**Comment by mohamedalyzahran** [comment]
sorry for my endless questions, but I got those messages while i am running the model however it seems to be working good
![image](https://github.com/sgoldenlab/simba/assets/123275892/66d48930-3176-4236-ba93-c4d3041fc462)

**Comment by sronilsson** [comment]
Yeah you don't have to worry, just a warning and never caused me any troubles. But if you to get rid of it:

Do `pip install scikit-learn==0.22.1` to install same version of scikit the classifier was created in, and start SimBA, I think.

no worries, people asking a lot more questions then this :) plus this helps me a lot to catch errors or oddeties.

**Comment by mohamedalyzahran** [comment]
Thanks so much. Now I am visualizing
And what you can see in the images that the program is mixing up the
animals I mean at the binning of the video it was ok (residend and
intruders are assigned correctly) but after sometime it is reversed  (the
resident assigned as intruder and the intruder assigned as resident)
Is this related to the video analysis by deeplabcut? If so how can I solve
this?
Or it is something in simba?

Thanks in advance

**Comment by sronilsson** [comment]
Hey @mohamedalyzahran - just so I understand - is this reversal happening across different videos, or is it happening within a video?

**Comment by mohamedalyzahran** [comment]
But I am not sure if the program will consider this change or mixing in
detecting the behavior or it ignore this mix

**Comment by mohamedalyzahran** [comment]
Within a single video it is mixing by time

**Comment by sronilsson** [comment]
OK, most likely a DLC thing: it can be difficult to tell in DLC visualizations sometimes, because (I think) by default the body-parts of the animals is drawn in the same color, left ear of animal 1 has same color as left ear of animal 2 etc. So you can't see the switches.

**Comment by mohamedalyzahran** [comment]
Actually the images that is attached were from the same video and it is
happening several times in the same video (changing the identity). Do you
know if there is a way to avoid this in dlc, could having a better model
(in term of training time, more videos, and more labeling) help

another question is that does simba measure the behavior (in each
classifier) taking into account this mixing up ie. When the identy change
within the video, simba detect the behavior of the other mouse or stick to
the same original one and ignore this mix?

**Comment by sronilsson** [comment]
Annotation, more videos and training help. I think the DLC people have set up some tutorials and documentation for how to get rid of ID switches, if you search for it on their forums you should find it. 

My bet is that a classifier like attack won't suffer too much, looking at mainly aggregate movement of both animals and their distances, but others will. Because, if the classifier thinks its being given a value and told its the velocity of the resident, and in fact its the velocity of the intruder, it will be confused. If you want other aggregate statistics (distance moved by animal etc) in simba, those will also mangled as the ID is not consistant across frames.

**Comment by mohamedalyzahran** [comment]
Do you recommend using another annotations software, does this gonna help
in switching identities

**Comment by sronilsson** [comment]
So yes more data is always good :), but there may also be further alternatives in DLC, how model is trained, a lot of people have this issue:

https://forum.image.sc/search?q=identity%20switches%20%23deeplabcut
